### ISSUE TITLE
Type checking not working for some functions

### ISSUE THREAD

**Issue by rkoeninger** [rkoeninger/ShenSharp#1]
(make sure you have `(tc +)` on)

These fail:

```
(define member {A --> (list A) --> boolean}
    _ [] -> false
    X [X | _] -> true
    X [_ | Y] -> (member X Y))
type error in rule 1 of member

(define never-contains? {A --> (list A) --> boolean}
    _ _ -> false)
type error in rule 1 of never-contains?

(define concat* {(list string) -> string}
    [] -> ""
    [X | Y] -> (cn X (concat* Y)))
type error in rule 1 of concat*

(define is-empty? {(list A) --> boolean}
    [] -> true
    [_ | _] -> false)
type error in rule 1 of is-empty?

(define all_Ns? {number --> (list number) --> boolean}
    _ [] -> true
    N [N | Ns] -> (all_Ns? N Ns)
    _ _ -> false)
type error in rule 1 of all_Ns?
```

These work:

```
(define identity {A --> A}
    X -> X)

(define factorial {number --> number}
    0 -> 1
    N -> (* N (factorial (- N 1))))

(define find-index {number --> A --> (list A) --> number}
    _ _ [] -> -1
    N X [X | _] -> N
    N X [_ | Y] -> (find-index (+ 1 N) X Y))

(define every-other {(list A) --> (list A)}
    [] -> []
    [X] -> [X]
    [X _ | Y] -> [X | (every-other Y)])

(define pairwise {(list A) --> (list (A * A))}
    [] -> []
    [_] -> (error "odd number of items in list")
    [X Y | Z] -> [(@p X Y) | (pairwise Z)])
```

**Comment by rkoeninger** [comment]
Looks like it's functions that return literal `true` and `false` that are having a problem. As if they are getting treated as symbols by the type checker even though `(symbol? true)` returns `false`.

These versions of the `member` function work:

```
(define member' {A --> (list A) --> boolean}
    _ [] -> (= 0 1)
    X [X | _] -> (= 0 0)
    X [_ | Y] -> (member' X Y))

(define member'' {A --> (list A) --> symbol}
    _ [] -> false
    X [X | _] -> true
    X [_ | Y] -> (member'' X Y))
```

**Comment by rkoeninger** [comment]
The Haskell implementation returns `true` for `(boolean? (intern "true"))`.

ShenSharp returns `false`.

**Comment by rkoeninger** [comment]
Above functions now work and `(boolean? (intern "true"))` now returns `true`.

Primitive Bool type was completely removed and the symbols `true` and `false` are used instead as was apparently intended.

Fixed in 8b8f999dbca4e92e85e931d652f599f7d3623f6c
